File: backend/app/schemas/boss.py
from typing import Optional, List
from pydantic import BaseModel, Field


class BossCreate(BaseModel):
    name: str
    title: str
    level: int = Field(ge=1, le=100, default=1)
    health: int = Field(ge=1, default=1000)
    max_health: int = Field(ge=1, default=1000)
    attack: int = Field(ge=0, default=50)
    defense: int = Field(ge=0, default=30)
    description: str
    lore: Optional[str] = None
    rewards: List[int] = Field(default_factory=list)
    image_url: Optional[str] = None
    icon_url: Optional[str] = None


class BossUpdate(BaseModel):
    name: Optional[str] = None
    title: Optional[str] = None
    level: Optional[int] = Field(None, ge=1, le=100)
    health: Optional[int] = Field(None, ge=1)
    max_health: Optional[int] = Field(None, ge=1)
    attack: Optional[int] = Field(None, ge=0)
    defense: Optional[int] = Field(None, ge=0)
    description: Optional[str] = None
    lore: Optional[str] = None
    rewards: Optional[List[int]] = None
    image_url: Optional[str] = None
    icon_url: Optional[str] = None

File: frontend/lib/data/models/boss.dart
class Boss {
  final int id;
  final String name;
  final String? title;
  final int level;
  final int health;
  final int maxHealth;
  final int attack;
  final int defense;
  final String? description;
  final String? lore;
  final List<int>? rewardIds;
  final String? imagePath;
  final String? iconPath;

  Boss({
    required this.id,
    required this.name,
    this.title,
    required this.level,
    required this.health,
    required this.maxHealth,
    required this.attack,
    required this.defense,
    this.description,
    this.lore,
    this.rewardIds,
    this.imagePath,
    this.iconPath,
  });

  factory Boss.fromJson(Map<String, dynamic> json) {
    return Boss(
      id: json['id'] as int,
      name: json['name'] as String,
      title: json['title'] as String?,
      level: json['level'] as int,
      health: json['health'] as int,
      maxHealth: json['max_health'] as int,
      attack: json['attack'] as int,
      defense: json['defense'] as int,
      description: json['description'] as String?,
      lore: json['lore'] as String?,
      rewardIds: json['reward_ids'] != null ? List<int>.from(json['reward_ids']) : null,
      imagePath: json['image_path'] as String?,
      iconPath: json['icon_path'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'title': title,
      'level': level,
      'health': health,
      'max_health': maxHealth,
      'attack': attack,
      'defense': defense,
      'description': description,
      'lore': lore,
      'reward_ids': rewardIds,
      'image_path': imagePath,
      'icon_path': iconPath,
    };
  }
}

